code/datums/components/bayonet_attachable.dm

```diff
@@ -0,0 +1,212 @@
+)
+
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.removable = removable
+	src.bayonet_icon_state = bayonet_icon_state
+	src.bayonet_overlay = bayonet_overlay
+	src.bayonet_overlay_icon = bayonet_overlay_icon
+	src.offset_x = offset_x
+	src.offset_y = offset_y
+	src.allow_sawnoff = allow_sawnoff
+
+	if (istype(starting_bayonet))
+		add_bayonet(starting_bayonet)
+
+/datum/component/bayonet_attachable/Destroy(force)
+	if(bayonet)
+		remove_bayonet()
+	return ..()
+
+/datum/component/bayonet_attachable/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_OBJ_DECONSTRUCT, PROC_REF(on_parent_deconstructed))
+	RegisterSignal(parent, COMSIG_ATOM_EXITED, PROC_REF(on_item_exit))
+	RegisterSignal(parent, COMSIG_ATOM_TOOL_ACT(TOOL_SCREWDRIVER), PROC_REF(on_screwdriver))
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_ICON_STATE, PROC_REF(on_update_icon_state))
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(on_update_overlays))
+	RegisterSignal(parent, COMSIG_ATOM_ATTACKBY, PROC_REF(on_attackby))
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_QDELETING, PROC_REF(on_parent_deleted))
+	RegisterSignal(parent, COMSIG_ITEM_PRE_ATTACK, PROC_REF(on_pre_attack))
+	RegisterSignal(parent, COMSIG_GUN_BEING_SAWNOFF, PROC_REF(on_being_sawnoff))
+	RegisterSignal(parent, COMSIG_GUN_SAWN_OFF, PROC_REF(on_sawn_off))
+
+/datum/component/bayonet_attachable/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_OBJ_DECONSTRUCT,
+		COMSIG_ATOM_EXITED,
+		COMSIG_ATOM_TOOL_ACT(TOOL_SCREWDRIVER),
+		COMSIG_ATOM_UPDATE_ICON_STATE,
+		COMSIG_ATOM_UPDATE_OVERLAYS,
+		COMSIG_ATOM_ATTACKBY,
+		COMSIG_ATOM_EXAMINE,
+		COMSIG_QDELETING,
+		COMSIG_ITEM_PRE_ATTACK,
+		COMSIG_GUN_BEING_SAWNOFF,
+		COMSIG_GUN_SAWN_OFF,
+	))
+
+/datum/component/bayonet_attachable/proc/on_examine(obj/item/source, mob/examiner, list/examine_list)
+	SIGNAL_HANDLER
+
+	if(!bayonet)
+		examine_list += "It has a <b>bayonet</b> lug on it."
+		return
+
+	examine_list += "It has \a [bayonet] [removable ? "" : "permanently "]affixed to it."
+	if(removable)
+		examine_list += span_info("[bayonet] looks like it can be <b>unscrewed</b> from [bayonet].")
+
+/datum/component/bayonet_attachable/proc/on_pre_attack(obj/item/source, atom/target, mob/living/user, params)
+	SIGNAL_HANDLER
+
+	if (!bayonet || !user.combat_mode)
+		return NONE
+
+	INVOKE_ASYNC(bayonet, TYPE_PROC_REF(/obj/item, melee_attack_chain), user, target, params)
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/datum/component/bayonet_attachable/proc/on_attackby(obj/item/source, obj/item/attacking_item, mob/attacker, params)
+	SIGNAL_HANDLER
+
+	if(!is_type_in_typecache(attacking_item, valid_bayonets))
+		return
+
+	if(bayonet)
+		source.balloon_alert(attacker, "already has \a [bayonet]!")
+		return
+
+	if(!attacker.transferItemToLoc(attacking_item, source))
+		return
+
+	add_bayonet(attacking_item, attacker)
+	source.balloon_alert(attacker, "attached [attacking_item]")
```

Comment:
```suggestion
	source.balloon_alert(attacker, "attached")
```
Avoid embedding names which can be an arbitrary length in balloon alerts
in this case the user always knows the item they were using, so we don't need to repeat it back to them